securityscanutils/github_writer.go

```diff
@@ -0,0 +1,127 @@
+package securityscanutils
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/google/go-github/v32/github"
+	"github.com/rotisserie/eris"
+	"github.com/solo-io/go-utils/githubutils"
+)
+
+type GithubRepo struct {
+	RepoName string
+	Owner    string
+}
+
+func (r GithubRepo) Address() string {
+	return fmt.Sprintf("github.com/%s/%s", r.Owner, r.RepoName)
+}
+
+// GithubIssueWriter is responsible for creating Github issues
+// to track vulnerabilities that have been discovered in images within a release.
+// It is configured with a Predicate that filters which releases to
+// write issues for, and which to skip
+type GithubIssueWriter struct {
+	// The details about the Github repository
+	repo GithubRepo
+
+	// The client used to write the issues
+	client *github.Client
+
+	// The RepositoryReleasePredicate used to determine if a vulnerability
+	// associated with a certain release should be uploaded to GitHub
+	createGithubIssuePredicate githubutils.RepositoryReleasePredicate
+
+	// A local cache of all existing GitHub issues
+	// Used to ensure that we are updating existing issues that were created by previous scans
+	allGithubIssues []*github.Issue
+}
+
+func NewGithubIssueWriter(repo GithubRepo, client *github.Client, issuePredicate githubutils.RepositoryReleasePredicate) *GithubIssueWriter {
+	return &GithubIssueWriter{
+		repo:                       repo,
+		client:                     client,
+		createGithubIssuePredicate: issuePredicate,
+		allGithubIssues:            nil, // initially nil, we'll lazy load these
+	}
+}
+
+// Labels that are applied to github issues that security scan generates
+var labels = []string{"trivy", "vulnerability"}
+
+// getAllGithubIssues returns the set of open issues in a Github repository that contain the trivy labels
+func (g *GithubIssueWriter) getAllGithubIssues(ctx context.Context) ([]*github.Issue, error) {
+	if g.allGithubIssues != nil {
+		// Maintain a local cache of issue to avoid re-requesting each time
+		return g.allGithubIssues, nil
+	}
+
+	issues, err := githubutils.GetAllIssues(ctx, g.client, g.repo.Owner, g.repo.RepoName, &github.IssueListByRepoOptions{
```

Comment:
This is dangerous as we shouldn't get all issues each time that we run for a given release as this is how we will get rate limited by github

Reply:
Actually looking further. Any error that we encounter when creating a github issue will bubble up and cause the job to fail. This is what the previous behavior was as well. 
I'm more than happy to improve this if we think it's valuable

Reply:
Oh nice I see your intermediate has a check